V5: Add Key-Manipulation Functions

brxken128 opened the pull request:
This adds the functionality to manipulate V5 header's keys (add/delete/change).

- [x] Use the `dexios key xx` subcommand
- [x] Change Key functionality
- [x] Key Add functionality
- [x] Key Remove functionality
- [x] ~~Update change key functionality to use the features added in #140~~ Not viable.
- [x] Give `header restore` the functionality to detect if there's empty space at the start of the file or not. If there's no empty space, disallow header restoration as `encrypted/detached` mode does not account for this. It's assumed the user knows what they are doing here, and I will make sure to add a note about it in the docs.
- [x] Add GA tests for all of the above
- [x] Make use of the newly exposed `vec_to_arr()`

pleshevskiy commented on dexios/src/subcommands/header_key.rs:
```diff
@@ -140,7 +142,112 @@ pub fn change_key(input: &str, key_old: &Key, key_new: &Key) -> Result<()> {
+            let master_key_encrypted = master_key_result
+                .map_err(|_| anyhow::anyhow!("Unable to encrypt your master key"))?;
+
+            let mut master_key_encrypted_array = [0u8; 48];
```
We already have `dexios_core::primitives::ENCRYPTED_MASTER_KEY_LEN;`

brxken128 replied:
A local commit updated to that, maybe I forgot to push :)

brxken128 commented:
@pleshevskiy I don't think `Update change key functionality to use the features added in https://github.com/brxken128/dexios/pull/140` is doable.

`decrypt_master_key()` doesn't return an index/the ciphertext/anything that matched, so I'm not able to replace it for `key change`, or use it for `key delete`. I could use it for `key add` though.

We'd need to return the index/matching ciphertext/anything to identify the keyslot we'd like to change or modify.

pleshevskiy commented:
each version of the headers is different) it is possible that this utility is not suitable for you) Don't worry) we'll get to make the core look nice and convenient, too)

brxken128 commented:
> each version of the headers is different) it is possible that this utility is not suitable for you) Don't worry) we'll get to make the core look nice and convenient, too)

Damn I am excited for that!

On another note, would you like collaborator access to this repo, and your name under the authors? I believe you've done more than enough - all that I ask is that changes are still done via PR, and that someone else has checked over them first - I'll stick to these rules too. Let me know and I'll send you the invite if so :)

pleshevskiy commented:
> On another note, would you like collaborator access to this repo, and your name under the authors? I believe you've done more than enough - all that I ask is that changes are still done via PR, and that someone else has checked over them first - I'll stick to these rules too. Let me know and I'll send you the invite if so :)

I would be very happy to work together on such a great project) I always make all changes through MR. Even in my small projects)

brxken128 commented:
> I would be very happy to work together on such a great project) I always make all changes through MR. Even in my small projects)

I've invited you! I try but sometimes I forget, but I'll adhere to those rules in future.

pleshevskiy commented:
Great! Now I can resolve conversation!

brxken128 commented:
@pleshevskiy How do we feel about exposing the `vec_to_arr()` function in the core?

I feel like we could get a lot more use out of it, even if we just define it in `key.rs` or make a new module called `utils.rs`.

brxken128 commented:
I originally added the change to allow environment variable specification of new keys, but I feel as though this will be awfully confusing to the user (with no way to differentiate between the two).

I reverted it, and then just decided to go with a second keyfile for these tests. It's more than sufficient, and easier to manage.

pleshevskiy commented on dexios/src/subcommands/header_key.rs:
```diff
@@ -140,7 +143,312 @@ pub fn change_key(input: &str, key_old: &Key, key_new: &Key) -> Result<()> {
-            todo!()
+            let mut keyslots = header.keyslots.clone().unwrap();
+
+            if keyslots.len() < 1 {
```
I suggest to use the following code:

```
let keyslot = keyslots.iter().next().ok_or_else(|| anyhow::anyhow!("No keyslots found, so we cannot continue."))?;
```

and then `keyslot` instead of `keyslots[0]`

brxken128 replied:
I like this change! I'll implement it now :)

pleshevskiy commented:
> How do we feel about exposing the vec_to_arr() function in the core?

I wanted to find out why the protected structure didn't like it.

![image](https://user-images.githubusercontent.com/7839004/180514114-e84a4b4a-2cff-4adb-a3a8-4be401c65eb1.png)

But we can remove Protected from this function and convert only vec to an array. Thus, we can use it as follows:

```
let vector = vec![];
let protected = Protected::new(vec_to_arr(vector));
```

brxken128 commented:
I updated what I could to the newly-exposed `vec_to_arr()`, although the master key decryption with the index could really do with being standardised. I'll take a proper look into it tomorrow - I don't think it'll be *too* hard.

pleshevskiy commented on dexios/src/subcommands/header_key.rs:
```diff
@@ -91,18 +92,14 @@ pub fn change_key(input: &str, key_old: &Key, key_new: &Key) -> Result<()> {
-            master_key_decrypted.zeroize();
-            let master_key = Protected::new(master_key);
+            let master_key =
+                Protected::<[u8; MASTER_KEY_LEN]>::new(vec_to_arr(master_key_decrypted));
```
Here we can write less annotations

```rust
Protected::new(vec_to_arr::<MASTER_KEY_LEN>(master_key_descrypted));
```

brxken128 replied:
Ah, I must have missed this one! I'll sort it when I can :)

brxken128 commented on dexios/src/subcommands/header.rs:
```diff
@@ -126,13 +126,39 @@ pub fn restore(input: &str, output: &str, skip: SkipMode) -> Result<()> {
+        if expected_empty_space.is_empty() {
+            header.write(&mut output_file)?;
+        } else {
+            return Err(anyhow::anyhow!("No empty space found at the start of {}! It's either not an encrypted file, or it was encrypted in detached mode (and the header can't be restored)", output));
```
@pleshevskiy I think this is a good compromise. Encrypting in detached mode removes the empty space from the start of the file (and makes restoring a lot harder), but *stripping* the header leaves that empty space intact, so we can restore. I'm going to tweak the code and such so that it's easier on the eyes :)

pleshevskiy replied:
But there is no more empty space in the new files encrypted by new detached mode) you will have an error) You can use [my solution](https://github.com/brxken128/dexios/blob/master/dexios/src/domain/decrypt.rs#L62)

brxken128 replied:
> But there is no more empty space in the new files encrypted by detached mode) you will have an error) You can use [my solution](https://github.com/brxken128/dexios/blob/master/dexios/src/domain/decrypt.rs#L62)

The point was to error :)

My plan is to entirely disallow restoring encrypted/detached headers entirely, but we can still restore `stripped` headers! Stripping just replaces the header with 0s, so the empty space is still present for us to write to.

If a user encrypts in detached mode, it's almost certain that they know what they're doing. There's more mystery surrounding how the files were encrypted and stuff, but they lose access to re-attaching the header to the file. We can add notes about this in the docs, and in the man page (when I get around to making one).

pleshevskiy replied:
As far as I know it will break on `read_exact` ) So my way is still better) and there you can call the error you want.)

brxken128 commented:
I plan to get #163 done, and maybe a `decrypt_master_key` variant that returns the index as well as the master key, so that this code can be a lot neater than it currently is.

The functionality is there and the tests pass, and my manual testing seems fine also.

brxken128 commented:
So I found a bug within the `domain/decrypt` area, and fixed it there as well as in the `header restore` utility. With this, I gave it an error type and fixed an incorrect error that I found.

I think this PR is good to go, but I'd be happy to make any changes.

pleshevskiy commented:
Looks good)
